pkg/clusteragent/admission/mutate/auto_instrumentation_test.go

```diff
@@ -964,9 +1075,157 @@ func TestInjectAutoInstrumentation(t *testing.T) {
+					Value: "false",
+				},
+			}...),
+			wantErr:     false,
+			setupConfig: func() { mockConfig.Set("apm_config.instrumentation.enabled", true) },
+		},
+		{
+			name: "Single Step Instrumentation: disable with label",
+			pod: fakePodWithParent(
+				"ns",
+				map[string]string{}, map[string]string{
+					"admission.datadoghq.com/enabled": "false",
+				}, []corev1.EnvVar{}, "replicaset", "test-deployment-123"),
+			expectedEnvs: []corev1.EnvVar{},
+			wantErr:      false,
+			setupConfig:  func() { mockConfig.Set("apm_config.instrumentation.enabled", true) },
+		},
+		{
+			name: "Single Step Instrumentation: default service name for ReplicaSet",
+			pod:  fakePodWithParent("ns", map[string]string{}, map[string]string{}, []corev1.EnvVar{}, "replicaset", "test-deployment-123"),
+			expectedEnvs: append(append(injectAllEnvs(), expBasicConfig()...), corev1.EnvVar{
+				Name:  "DD_SERVICE",
+				Value: "test-deployment",
+			}),
+			wantErr:     false,
+			setupConfig: func() { mockConfig.Set("apm_config.instrumentation.enabled", true) },
+		},
+		{
+			name: "Single Step Instrumentation: default service name for StatefulSet",
+			pod:  fakePodWithParent("ns", map[string]string{}, map[string]string{}, []corev1.EnvVar{}, "statefulset", "test-statefulset-123"),
+			expectedEnvs: append(append(injectAllEnvs(), expBasicConfig()...), corev1.EnvVar{
+				Name:  "DD_SERVICE",
+				Value: "test-statefulset-123",
+			}),
+			wantErr:     false,
+			setupConfig: func() { mockConfig.Set("apm_config.instrumentation.enabled", true) },
+		},
+		{
+			name:         "Single Step Instrumentation: default service name (disabled)",
+			pod:          fakePodWithParent("ns", map[string]string{}, map[string]string{}, []corev1.EnvVar{}, "replicaset", "test-deployment-123"),
+			expectedEnvs: []corev1.EnvVar{},
+			wantErr:      false,
+			setupConfig:  func() { mockConfig.Set("apm_config.instrumentation.enabled", false) },
+		},
+		{
+			name:         "Single Step Instrumentation: disabled namespaces should not be instrumented",
+			pod:          fakePodWithParent("ns", map[string]string{}, map[string]string{}, []corev1.EnvVar{}, "replicaset", "test-app-123"),
+			expectedEnvs: []corev1.EnvVar{},
+			wantErr:      false,
+			setupConfig: func() {
+				mockConfig.Set("apm_config.instrumentation.enabled", true)
+				mockConfig.Set("apm_config.instrumentation.disabled_namespaces", []string{"ns"})
+			},
+		},
+		{
+			name: "Single Step Instrumentation: enabled namespaces should be instrumented",
+			pod:  fakePodWithParent("ns", map[string]string{}, map[string]string{}, []corev1.EnvVar{}, "replicaset", "test-app-123"),
+			expectedEnvs: append(append(injectAllEnvs(), expBasicConfig()...), corev1.EnvVar{
+				Name:  "DD_SERVICE",
+				Value: "test-app",
+			}),
+			wantErr: false,
+			setupConfig: func() {
+				mockConfig.Set("apm_config.instrumentation.enabled", false)
+				mockConfig.Set("apm_config.instrumentation.enabled_namespaces", []string{"ns"})
+			},
+		},
```

Comment:
I think we should add test cases for the UST tags `tags.datadoghq.com/{env, service, version}` and also make sure that the if the UST service tag is set that we do not override it with the deployment name (user value takes precedence).

Reply:
We cannot test this logic in this unit-test. Tags injection and library injection are different endpoints of the Admission Controller https://github.com/DataDog/datadog-agent/blob/main/cmd/cluster-agent/subcommands/start/command.go#L327-L329. I can add a test for this use case to the system-tests, but for now tested manually and the UST is respected.